lib/streamlit/git_util.py

```diff
@@ -0,0 +1,82 @@
+import git  # type: ignore[import]
+import os
+import re
+
+# Github has two URLs, one that is https and one that is ssh
+GITHUB_HTTP_URL = r"^https://(www\.)?github.com/(.+)/(.+).git$"
+GITHUB_SSH_URL = r"^git@github.com:(.+)/(.+).git$"
+
+
+class GitRepo:
+    def __init__(self, path):
+        try:
+            self.repo = git.Repo(path, search_parent_directories=True)
+            git_root = self.repo.git.rev_parse("--show-toplevel")
+            self.module = os.path.relpath(path, git_root)
+        except:
+            self.repo = None
+
+    def is_valid(self):
+        return self.repo is not None
+
+    def get_tracking_branch(self):
```

Comment:
For bonus Pythonic points, this kind of code could be made into a property:
```
@property
def tracking_branch(self):
...
# Later:
self.tracking_branch # No parenthesis!
```

Reply:
I can look into that.